src/wallet/wallet2.cpp

```diff
@@ -7761,6 +7761,194 @@ bool wallet2::check_tx_proof(const crypto::hash &txid, const cryptonote::account
+
+  total = spent = 0;
+  auto tx_as_hex = gettx_res.txs_as_hex.begin();
+  for (size_t i = 0; i < proofs.size(); ++i)
+  {
+    cryptonote::blobdata tx_data;
+    if (gettx_res.txs.size() == proofs.size())
+      ok = string_tools::parse_hexstr_to_binbuff(gettx_res.txs[i].as_hex, tx_data);
+    else
+      ok = string_tools::parse_hexstr_to_binbuff(*tx_as_hex++, tx_data);
+    THROW_WALLET_EXCEPTION_IF(!ok, error::wallet_internal_error, "Failed to parse transaction from daemon");
+
+    crypto::hash tx_hash, tx_prefix_hash;
+    cryptonote::transaction tx;
+    THROW_WALLET_EXCEPTION_IF(!cryptonote::parse_and_validate_tx_from_blob(tx_data, tx, tx_hash, tx_prefix_hash), error::wallet_internal_error,
+      "Failed to validate transaction from daemon");
+    THROW_WALLET_EXCEPTION_IF(tx_hash != proofs[i].txid, error::wallet_internal_error, "Failed to get the right transaction from daemon");
+
+    bool found = false;
+    for (size_t n = 0; n < tx.vout.size(); ++n)
+    {
+      const cryptonote::txout_to_key* const out_key = boost::get<cryptonote::txout_to_key>(std::addressof(tx.vout[n].target));
+      if (!out_key || proofs[i].key != out_key->key)
+        continue;
+      found = true;
+
+      std::string prefix_data((const char*)&proofs[i].txid, sizeof(crypto::hash));
+      prefix_data += std::string((const char*)&proofs[i].key, sizeof(crypto::public_key));
+      prefix_data += message;
+      crypto::hash prefix_hash;
+      crypto::cn_fast_hash(prefix_data.data(), prefix_data.size(), prefix_hash);
+
+      if (tx.version == 1)
+      {
+        THROW_WALLET_EXCEPTION_IF(proofs[i].amount != tx.vout[n].amount, error::wallet_internal_error,
+          "The amount in the signature doesn't agree with the amount stored in the blockchain");
+        const std::vector<const crypto::public_key*> pubs = { &proofs[i].key };
+        if (!crypto::check_ring_signature(prefix_hash, proofs[i].ki, &pubs[0], 1, &proofs[i].nonrctSig))
+          return false;
+      }
+      else
+      {
+        rct::ctkeyV pubs(1);
+        pubs[0].dest = rct::pk2rct(proofs[i].key);
+        if (tx.rct_signatures.type == rct::RCTTypeNull)
+        {
+          pubs[0].mask = rct::zeroCommit(proofs[i].amount);
+        }
+        else
+        {
+          pubs[0].mask = tx.rct_signatures.outPk[n].mask;
+        }
+        const rct::key C = rct::scalarmultH(rct::d2h(proofs[i].amount));
+        proofs[i].rctSig.II = { rct::ki2rct(proofs[i].ki) };              // because mgSig doesn't serialize key image
+        if (!rct::verRctMGSimple(rct::hash2rct(prefix_hash), proofs[i].rctSig, pubs, C))
+          return false;
+      }
+      total += proofs[i].amount;
+      if (kispent_res.spent_status[i])
+        spent += proofs[i].amount;
+      break;
+    }
+    THROW_WALLET_EXCEPTION_IF(!found, error::wallet_internal_error, "The output wasn't found in the specified tx");
+  }
+
+  return true;
+}
```

Comment:
This doesn't seem to attach an address to the proof. Is this intended ? I suppose it can be done via the message anyway.

Reply:
Yes, it's intentional. Let me summarize these different proof schemes which are the following three:

1. Tx proof,  which proves: "I know the tx secret key of this `<txid>` where this much was sent to that `<address>`".
2. Spend proof, which proves: "I am the author of this `<txid>` hence I can recreate a new set of ring signatures consisting of the exact same set of decoy outputs (I can do so because I know the secret key of the real spent output of each ring)."
3. Reserve proof, which proves: "I know the secret key of this output key whose key image is that. Also its Pedersen Commitment commits to this amount."

As you can see, the address is part of the proof for 1, but not for 2 & 3. Hence 2 & 3 don't take address as input.